### PR TITLE
Persist our canvas state to Google Realtime API

### PR DESCRIPTION
**Pull request by girum11**

### CODE REVIEW

File: src/client/app/graph/canvas/canvas.component.html [hunk 1]
```diff
@@ -4,5 +4,7 @@
      xmlns:xlink="http://www.w3.org/1999/xlink"
      (mousedown)="this.onMousedown($event)"
      (mousewheel)="this.onMousewheel($event)">
-  <g uxg-card *ngFor="let card of this.canvasService.cards" [card]="card"></g>
+  <g uxg-card
+     *ngFor="let card of (this.canvasService.cards2 | toIterable)"
```
Comment (nati22) [inline src/client/app/graph/canvas/canvas.component.html]: cards2?

**Comment by girum11** [comment]
Yeah the PR isn't ready yet, but I wanted to see the net files changed
across my commits.

Once I clean it out to the point that it's ready for review, I'll comment
to let you guys know.



In the meantime, feel free to check it out though.

**Comment by nati22** [comment]
Sounds good. @eyuelt and I will keep lurking

**Comment by eyuelt** [comment]
For future reference, you can do this without making a PR. Just add /compare to the repo url, https://github.com/tessa3/uxgraph/compare, and enter the branches or commits you want to compare. 

Example: https://github.com/tessa3/uxgraph/compare/master...collaborative-state

**Comment by girum11** [comment]
Ahh Eyuel how do you understand both Critique and GitHub?? They're both so
confusing 😶

**Comment by girum11** [comment]
Okay, this PR should be ready for review now.

We can go over it together next time we meet up so that I can really walk it through.

**Comment by nati22** [comment]
Yeah I think a walkthrough would be helpful for this PR in particular

**Comment by girum11** [comment]
This PR will close #124, #125 and #129

File: src/client/app/graph/canvas/canvas.component.ts [hunk 2]
```diff
@@ -106,4 +106,8 @@ export class CanvasComponent {
       this.canvasService.multiSelectMode = false;
     }
   }
+
+  ngDoCheck(): void {
```
Comment (eyuelt) [inline src/client/app/graph/canvas/canvas.component.ts]: Remove this

File: src/client/app/graph/card/card.component.ts [hunk 3]
```diff
@@ -16,7 +19,12 @@ import {EventUtils} from '../../utils/event-utils';
 })
 export class CardComponent implements OnInit {
   // The card data to render on the canvas.
-  @Input() card: Card = null;
+  // TODO(girum): Give these realtime custom models real static types.
+  @Input() card: any = null;
+
+  // xPosition: number;
```
Comment (eyuelt) [inline src/client/app/graph/card/card.component.ts]: Remove these comments?

File: src/client/app/graph/card/card.component.ts [hunk 4]
```diff
@@ -30,12 +38,17 @@ export class CardComponent implements OnInit {
...
+  constructor(private canvasService: CanvasService,
+              private googleRealtimeService: GoogleRealtimeService) {
   }
 
   ngOnInit() {
     this.position = this.canvasService.canvasCoordToViewportCoord(this.card);
     this.canvasService.addListener(this.update.bind(this));
+    this.googleRealtimeService.currentDocument.subscribe(document => {
+      document.getModel().getRoot()
+          .addEventListener(OBJECT_CHANGED, this.update.bind(this));
```
Comment (eyuelt) [inline src/client/app/graph/card/card.component.ts]: Add a TODO to change this to only listen to events on this card, rather than on the whole document.

**Pull request merged**